src/storage/QueryBaseProcessor.inl

```diff
@@ -255,16 +254,66 @@ kvstore::ResultCode QueryBaseProcessor<REQ, RESP>::collectEdgeProps(
     return ret;
 }
 
+template<typename REQ, typename RESP>
+folly::Future<std::vector<OneVertexResp>>
+QueryBaseProcessor<REQ, RESP>::asyncProcessBucket(Bucket bucket) {
+    folly::Promise<std::vector<OneVertexResp>> pro;
+    auto f = pro.getFuture();
+    executor_->add([this, p = std::move(pro), b = std::move(bucket)] () mutable {
+        std::vector<OneVertexResp> codes;
+        codes.reserve(b.vertices_.size());
+        for (auto& pv : b.vertices_) {
+            codes.emplace_back(pv.first, pv.second, processVertex(pv.first, pv.second));
+        }
+        p.setValue(std::move(codes));
+    });
+    return f;
+}
+
+template<typename REQ, typename RESP>
+int32_t QueryBaseProcessor<REQ, RESP>::getBucketsNum(int32_t verticesNum, int handlerNum) {
+    if (verticesNum < FLAGS_min_vertices_num_mp) {
+        return 1;
+    }
+    return std::min(verticesNum, handlerNum);
+}
+
+template<typename REQ, typename RESP>
+std::vector<Bucket> QueryBaseProcessor<REQ, RESP>::genBuckets(
```

Comment:
the benchmark and ut do not directly cover different vertexnum and max_handlers_per_req. 
the code looks fine. add a ut is better.

Reply:
Good point.  I will add more factors for genBuckets method,  for example,  current io load, cpu usages etc.

Reply:
this requires more performance counters. a TODO is enough now.

Reply:
Fair enough!

Reply:
I have added some UTs for genBucket method.